nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/PasswordBasedOauth2TokenProvider.java

```diff
@@ -0,0 +1,230 @@
+        .name("client-secret")
+        .displayName("Client secret")
+        .dependsOn(CLIENT_ID)
+        .required(true)
+        .sensitive(true)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .build();
+
+    public static final PropertyDescriptor SSL_CONTEXT = new PropertyDescriptor.Builder()
+        .name("ssl-context")
+        .displayName("SSL Context")
+        .addValidator(Validator.VALID)
+        .identifiesControllerService(SSLContextService.class)
+        .required(false)
+        .build();
+
+    private static final List<PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(Arrays.asList(
+        AUTHORIZATION_SERVER_URL,
+        USERNAME,
+        PASSWORD,
+        CLIENT_ID,
+        CLIENT_SECRET,
+        SSL_CONTEXT
+    ));
+
+    private String authorizationServerUrl;
+    private OkHttpClient httpClient;
+
+    private String username;
+    private String password;
+    private String clientId;
+    private String clientSecret;
+
+    private volatile AccessDetails accessDetails;
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        authorizationServerUrl = context.getProperty(AUTHORIZATION_SERVER_URL).evaluateAttributeExpressions().getValue();
+
+        httpClient = createHttpClient(context);
+
+        username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+        password = context.getProperty(PASSWORD).getValue();
+        clientId = context.getProperty(CLIENT_ID).evaluateAttributeExpressions().getValue();
+        clientSecret = context.getProperty(CLIENT_SECRET).getValue();
+    }
+
+    protected OkHttpClient createHttpClient(ConfigurationContext context) {
+        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();
+
+        SSLContextService sslService = context.getProperty(SSL_CONTEXT).asControllerService(SSLContextService.class);
+        if (sslService != null) {
+            final X509TrustManager trustManager = sslService.createTrustManager();
+            SSLContext sslContext = sslService.createContext();
+            clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), trustManager);
+        }
+
+        return clientBuilder.build();
+    }
+
+    @Override
+    public String getAccessToken() {
+        String accessToken;
+
+        if (this.accessDetails == null) {
+            acquireAuthorizationDetails();
+        } else if (this.accessDetails.isExpired()) {
+            try {
+                refreshAuthorizationDetails();
+            } catch (Exception e) {
+                getLogger().info("Couldn't refresh access token, probably refresh token has expired." +
+                    " Getting new token using credentials.");
+                acquireAuthorizationDetails();
```

Comment:
This method could also throw any number of RuntimeExceptions, are expecting the caller to implement optional handling of RuntimeExceptions?

Reply:
I don't see this any more special than other methods (which "could also throw any number of RuntimeExceptions"). The framework-provided default error-handling should suffice in my opinion.

Reply:
As mentioned in the other comment, breaking out exception handling between issues with HTTP responses, versus other types of issues would help determine whether to call `acquireAuthorizationDetails()`. At that point, I agree that framework-level handling is sufficient for exception processing.

Reply:
I tested all error scenarios and in all cases I see the issue properly reported.
The information provided is clear and lacks no detail.
Unless presented a concrete detailed and reproducible example that would prove otherwise I consider this level of error handling sufficient.

Reply:
Thanks for the response, I added a comment with some additional details and one potential suggestion.